.github/actions/browser/action.yml

```diff
@@ -0,0 +1,82 @@
+name: 'Run Test On Browser'
+description: 'Setup Browser and Run Test'
+inputs:
+  # Support one browser per action for now, because some packages has build script before test
+  # which nx cannot send params to karma test after build
+  browsers:
+    description: 'Browser name'
+    required: false
+    default: 'default'
+  mode:
+    description: 'Test mode'
+    required: false
+    default: 'affected' # affected|all
+  coverage:
+    description: 'Include test covertage'
+    required: false
+    default: 'true'
+  parallel:
+    description: 'Run parallel testing limit'
+    required: false
+    # Default value 3 is OK for every browsers except Safari need to be one to make it stable
+    default: 1
+  base:
+    description: 'Base sha for test affected mode'
+    required: false
+  browserstack:
+    description: 'Run test on BrowserStack'
+    required: false
+    default: 'false'
+  username:
+    description: 'BrowserStack username'
+    required: false
+  access-key:
+    description: 'BrowserStack access key'
+    required: false
+runs:
+  using: "composite"
+  steps:
+    - name: 'BrowserStack Env Setup' # Invokes the setup-env action
+      if: ${{ inputs.browserstack == 'true' }}
+      uses: browserstack/github-actions/setup-env@master
+      with:
+        username: ${{ inputs.username }}
+        access-key: ${{ inputs.access-key }}
+
+    - name: 'BrowserStack Local Tunnel Setup' # Invokes the setup-local action
+      if: ${{ inputs.browserstack == 'true' }}
+      uses: browserstack/github-actions/setup-local@master
+      with:
+        local-testing: start
+        local-identifier: random
+
+    - name: Setup resources and environment
+      uses: ./.github/actions/setup
+      id: setup
+
+    - name: Setup browsers or browserstack command line option
+      shell:
+      run: |
+        [[ "${{ inputs.browserstack }}" == "true" ]] && export BROWSER=browserstack || export BROWSER=browsers
+        echo "BROWSERS_OPTION=--$BROWSER='${{ inputs.browsers }}'" >> $GITHUB_ENV
+
+    # Workaround fix NX cannot send param to Karma CLI in some packages
+    - name: Update browsers option
+      shell:
+      run: |
+        [[ "${{ inputs.browserstack }}" == "false" && "${{ inputs.browsers }}" == "default" ]] && echo "BROWSERS_OPTION=" >> $GITHUB_ENV
+        echo "Updated"
+
+    - name: Test
+      shell: bash
+      # Prevent NX caching for mode `all` because it must run all testing, even though the code has not chaged.
```

Comment:
Why should it always run the test even the code does not change ?

Reply:
We do not need to cache the test. when run nightly testing.